crates/core/tedge_agent/src/file_transfer_server/error.rs

```diff
@@ -24,8 +23,85 @@ pub enum FileTransferError {
     BindingAddressInUse { address: std::net::SocketAddr },
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum FileTransferRequestError {
+    #[error(transparent)]
+    FromIo(#[from] std::io::Error),
+
+    #[error("Request to delete {path:?} failed: {err}")]
+    DeleteIoError {
+        #[source]
+        err: std::io::Error,
+        path: RequestPath,
+    },
+
+    #[error("Request to upload to {path:?} failed: {err:?}")]
+    Upload {
+        #[source]
+        err: anyhow::Error,
+        path: RequestPath,
+    },
+
+    #[error("Invalid file path: {path:?}")]
+    InvalidPath { path: RequestPath },
+
+    #[error("File not found: {0:?}")]
+    FileNotFound(RequestPath),
+
+    #[error("Path rejection: {0:?}")]
+    PathRejection(#[from] PathRejection),
+}
+
 impl From<FileTransferError> for RuntimeError {
     fn from(error: FileTransferError) -> Self {
         RuntimeError::ActorError(Box::new(error))
     }
 }
+
+impl IntoResponse for FileTransferError {
+    fn into_response(self) -> axum::response::Response {
+        use FileTransferError::*;
+        let status_code = match self {
+            FromIo(_)
+            | FromHyperError(_)
+            | FromAddressParseError(_)
+            | FromUtf8Error(_)
+            | BindingAddressInUse { .. } => {
+                tracing::error!("{self}");
+                StatusCode::INTERNAL_SERVER_ERROR
+            }
+        };
+        status_code.into_response()
+    }
+}
+
+impl IntoResponse for FileTransferRequestError {
+    fn into_response(self) -> axum::response::Response {
+        use FileTransferRequestError::*;
+        match &self {
+            FromIo(_) | PathRejection(_) => {
+                tracing::error!("{self}");
```

Comment:
I'm not sure axum's `tracing` is quite what we want here. Here, I want to determine whether I'm hiding information from the user (which is when we have an internal server error), and if so, log that information. This avoids polluting the server logs with errors arising from clients misusing the API, but means we don't expose implementation details to the user of the API. If we rely on some sort of middleware for logging, either it would take an input of `FileTransferRequestError`, in which case we might as well do everything here as we'd still need to log selectively, or we'd take the converted response, by which point we know the status code, so it's easy to log selectively, but we've thrown away the pertinent information (e.g. an IO error).